File: 0072-edit-distance/0072-edit-distance.ts
function minDistance(word1: string, word2: string): number {
    let dp = Array.from({length:word1.length+1},()=>Array(word2.length+1).fill(-1))
    return helper(word1.length-1,word2.length-1,word1,word2,dp)
};

function helper(i,j,word1,word2,dp){
    if (i == -1){
        return j + 1
    }
    if (j == -1){
        return i + 1
    }
    if ( dp[i][j] != -1 ) {
        return dp[i][j]
    }
    if (word1[i] == word2[j]){
       return dp[i][j] = helper(i - 1, j - 1, word1, word2, dp);
    }
    let insertion =  helper(i-1,j,word1,word2,dp)
    let deletion =  helper(i,j-1,word1,word2,dp)
    let replace = helper(i-1,j-1,word1,word2,dp)
    
    return dp[i][j] = 1 + Math.min(insertion,deletion,replace)

}

File: 0072-edit-distance/0072-edit-distance.java
class Solution {
    public int minDistance(String word1, String word2) {
        int l1 = word1.length();
        int l2 = word2.length();
        int[][] costDP = new int[l1+1][l2+1];
        for (int i = 0; i < l1+1; i++){
            costDP[i][0] = i;
        }
        for (int i = 0; i < l2+1; i++){
            costDP[0][i] = i;
        }
        for (int i = 1; i < l1+1; i++){
            for (int j = 1; j < l2+1; j++){
                if (word1.charAt(i-1)==word2.charAt(j-1)){
                    costDP[i][j] = costDP[i-1][j-1];
                }else {
                int topLeft = costDP[i-1][j-1];
                int top = costDP[i-1][j];
                int left = costDP[i][j-1];
                costDP[i][j] = Math.min(topLeft,Math.min(top,left))+1;
            }      
            } 
        }
        return costDP[l1][l2];
    }
}